README.md
# README: SER 421 Lab 2 pkedilay

To grade, use the code in the Extra Credits folder for the required functionality with Extra Credits functionalities.

If something fails, use code in this (root) folder.

## C4: How to instantiate

1. How to instantiate newsStore object:

    `const newsStore = require('./NewsService');`

    `const store = new newsStore.newsStore();`
2. Invoking APIs

    `newsStoreObj.writeNewNewsStory('Parikshith', 'Software Engineering', true, 'This is a sample work', new Date());`

    `newsStoreObj.updateHeadlineOfNewsStory(0, 'Web Development');`

    `newsStoreObj.changeContentOfExistingNewsStory(0, 'Dr. Gary', 'Socket Programming', false, 'Networking in Java', new Date(2018, 07, 01));`

    `newsStoreObj.deleteAnExistingNewsStory(0);`

    `newsStoreObj.getNewsByFilter('Web Development', 'Parikshith', new Date(1999, 01, 01), new Date(1999, 10, 10));`

## To Test Activity 1

I have used module exports, so the code can be tested directly by running

`node NewsService_test.js`

Instructions to run `node .load` way for `NewsService.js` is written in the file itself.

## To Test Activity 2

1. Run the following command to start the server:

    `node activity2.js`
2. Then go to `http://localhost:3000` on your web browser to access the application

3. Use these user details to login:
    1. Username: `guest` Password: `guest` Role: `Guest`
    2. Username: `subscriber` Password: `subscriber` Role: `Subscriber`
    3. Username: `author` Password: `author` Role: `Author`
    4. Username: `author1` Password: `author1` Role: `Author`
    5. Username: `author2` Password: `author2` Role: `Author`

## Some Important Information

1. Logs are written once every 5 seconds into `activity2.log` file
2. To test Activity 1 using test file, `news.txt` must be empty. This is done to get persistent results, so first two lines of code in
`ExtraCredits` > `NewsService_test.js` has been added to flush out any old data in that file.


ExtraCredits/news.txt
{"1":{"id":1,"author":"Parikshith","title":"Software Engineering","isPublic":true,"storyContent":"This is a sample work","date":"2000-02-01T07:00:00.000Z"},"2":{"id":2,"author":"Parikshith","title":"Web Development","isPublic":true,"storyContent":"This is a sample work","date":"2020-02-01T07:00:00.000Z"},"3":{"id":3,"author":"Kedilaya","title":"Software Engineering","isPublic":true,"storyContent":"This is a sample work","date":"2019-02-01T07:00:00.000Z"},"4":{"id":4,"author":"Kedilaya","title":"Web Development","isPublic":true,"storyContent":"This is a sample work","date":"2018-02-01T07:00:00.000Z"}}

NewsService_test.js
var newsService = require('./NewsService');
var newsStoreObj = new newsService.newsStore();
// If using .load, comment the above two lines and uncomment below line
// var newsStoreObj = new newsStore();

var allTestCasesPassed = true;

function testWriteNewNewsStory() {
    newsStoreObj.writeNewNewsStory('Parikshith', 'Software Engineering', true, 'This is a sample work', new Date());
    assertEquals(Object.keys(newsStoreObj.news).length, 1);
}

function testUpdateHeadlineOfNewsStory() {
    newsStoreObj.updateHeadlineOfNewsStory(0, 'Web Development');
    assertEquals(newsStoreObj.news[0].title, 'Web Development');
}

function testChangeContentOfExistingNewsStory() {
    newsStoreObj.changeContentOfExistingNewsStory(0, 'Dr. Gary', 'Socket Programming', false, 'Networking in Java', new Date(2018, 07, 01));
    assertEquals(newsStoreObj.news[0].author, 'Dr. Gary');
}

function testDeleteAnExistingNewsStory() {
    newsStoreObj.deleteAnExistingNewsStory(0);
    assertEquals(newsStoreObj.news[0]);
}

function testDeleteAnExistingNewsStoryIfItDoesnotExist() {
    try {
        newsStoreObj.deleteAnExistingNewsStory(10);
        console.log('Failed');
        allTestCasesPassed = false;
    } catch (e) {
        console.log('Passed');
    }
}

function generateDataForFilterTestCases() {
    newsStoreObj.writeNewNewsStory('Parikshith', 'Software Engineering', true, 'This is a sample work', new Date(2000, 01, 01));
    newsStoreObj.writeNewNewsStory('Parikshith', 'Web Development', true, 'This is a sample work', new Date(2020, 01, 01));
    newsStoreObj.writeNewNewsStory('Kedilaya', 'Software Engineering', true, 'This is a sample work', new Date(2019, 01, 01));
    newsStoreObj.writeNewNewsStory('Kedilaya', 'Web Development', true, 'This is a sample work', new Date(2018, 01, 01));
}

function testGetNewsByFilterTitle() {
    let filteredResults = newsStoreObj.getNewsByFilter('Software Engineering');
    assertEquals(filteredResults[0].date.getTime(), new Date(2000, 01, 01).getTime());
    assertEquals(filteredResults[1].date.getTime(), new Date(2019, 01, 01).getTime());
}

function testGetNewsByFilterAuthor() {
    let filteredResults = newsStoreObj.getNewsByFilter(undefined, 'Parikshith');
    assertEquals(filteredResults[0].date.getTime(), new Date(2000, 01, 01).getTime());
    assertEquals(filteredResults[1].date.getTime(), new Date(2020, 01, 01).getTime());
}

function testGetNewsByFilterRangeDate() {
    let filteredResults = newsStoreObj.getNewsByFilter(undefined, undefined, new Date(2018, 01, 01), new Date(2021, 01, 01));
    assertEquals(filteredResults.length, 3);
}

function testGetNewsByFilterTitleAuthor() {
    let filteredResults = newsStoreObj.getNewsByFilter('Web Development', 'Kedilaya');
    assertEquals(filteredResults.length, 1);
}

function testGetNewsByFilterTitleAuthorRangeDate() {
    let filteredResults = newsStoreObj.getNewsByFilter('Web Development', 'Parikshith', new Date(1999, 01, 01), new Date(1999, 10, 10));
    assertEquals(filteredResults.length, 0);
}

function assertEquals(actual, expected) {
    if (actual === expected) {
        console.log('Passed');
    } else {
        console.log('Failed');
        allTestCasesPassed = false;
    }
}

function driverTestCases() {
    testWriteNewNewsStory();
    testUpdateHeadlineOfNewsStory();
    testChangeContentOfExistingNewsStory();
    testDeleteAnExistingNewsStory();
    testDeleteAnExistingNewsStoryIfItDoesnotExist();

    // IMPORTANT: Do this before calling filter test cases
    generateDataForFilterTestCases();

    testGetNewsByFilterTitle();
    testGetNewsByFilterAuthor();
    testGetNewsByFilterRangeDate();
    testGetNewsByFilterTitleAuthor();
    testGetNewsByFilterTitleAuthorRangeDate();

    if (allTestCasesPassed) console.log("All test cases passed");
    else console.log("Some test cases Failed!");
}

// Starts here!
driverTestCases();



ExtraCredits/NewsService_test.js
//Adding these two lines to flush out any old data present in the news.txt file
var fs = require('fs');
fs.writeFileSync('./news.txt', '', {encoding:'utf8', flag:'w'});

var newsService = require('./NewsService');
var newsStoreObj = new newsService.newsStore();
// If using .load, comment the above two lines and uncomment below line
// var newsStoreObj = new newsStore();

var allTestCasesPassed = true;

function testWriteNewNewsStory() {
    newsStoreObj.writeNewNewsStory('Parikshith', 'Software Engineering', true, 'This is a sample work', new Date());
    assertEquals(Object.keys(newsStoreObj.news).length, 1);
}

function testUpdateHeadlineOfNewsStory() {
    newsStoreObj.updateHeadlineOfNewsStory(0, 'Web Development');
    assertEquals(newsStoreObj.news[0].title, 'Web Development');
}

function testChangeContentOfExistingNewsStory() {
    newsStoreObj.changeContentOfExistingNewsStory(0, 'Dr. Gary', 'Socket Programming', false, 'Networking in Java', new Date(2018, 07, 01));
    assertEquals(newsStoreObj.news[0].author, 'Dr. Gary');
}

function testDeleteAnExistingNewsStory() {
    newsStoreObj.deleteAnExistingNewsStory(0);
    assertEquals(newsStoreObj.news[0]);
}

function testDeleteAnExistingNewsStoryIfItDoesnotExist() {
    try {
        newsStoreObj.deleteAnExistingNewsStory(10);
        console.log('Failed');
        allTestCasesPassed = false;
    } catch (e) {
        console.log('Passed');
    }
}

function generateDataForFilterTestCases() {
    newsStoreObj.writeNewNewsStory('Parikshith', 'Software Engineering', true, 'This is a sample work', new Date(2000, 01, 01));
    newsStoreObj.writeNewNewsStory('Parikshith', 'Web Development', true, 'This is a sample work', new Date(2020, 01, 01));
    newsStoreObj.writeNewNewsStory('Kedilaya', 'Software Engineering', true, 'This is a sample work', new Date(2019, 01, 01));
    newsStoreObj.writeNewNewsStory('Kedilaya', 'Web Development', true, 'This is a sample work', new Date(2018, 01, 01));
}

function testGetNewsByFilterTitle() {
    let filteredResults = newsStoreObj.getNewsByFilter('Software Engineering');
    assertEquals(filteredResults[0].date.getTime(), new Date(2000, 01, 01).getTime());
    assertEquals(filteredResults[1].date.getTime(), new Date(2019, 01, 01).getTime());
}

function testGetNewsByFilterAuthor() {
    let filteredResults = newsStoreObj.getNewsByFilter(undefined, 'Parikshith');
    assertEquals(filteredResults[0].date.getTime(), new Date(2000, 01, 01).getTime());
    assertEquals(filteredResults[1].date.getTime(), new Date(2020, 01, 01).getTime());
}

function testGetNewsByFilterRangeDate() {
    let filteredResults = newsStoreObj.getNewsByFilter(undefined, undefined, new Date(2018, 01, 01), new Date(2021, 01, 01));
    assertEquals(filteredResults.length, 3);
}

function testGetNewsByFilterTitleAuthor() {
    let filteredResults = newsStoreObj.getNewsByFilter('Web Development', 'Kedilaya');
    assertEquals(filteredResults.length, 1);
}

function testGetNewsByFilterTitleAuthorRangeDate() {
    let filteredResults = newsStoreObj.getNewsByFilter('Web Development', 'Parikshith', new Date(1999, 01, 01), new Date(1999, 10, 10));
    assertEquals(filteredResults.length, 0);
}

function assertEquals(actual, expected) {
    if (actual === expected) {
        console.log('Passed');
    } else {
        console.log('Failed');
        allTestCasesPassed = false;
    }
}

function driverTestCases() {
    testWriteNewNewsStory();
    testUpdateHeadlineOfNewsStory();
    testChangeContentOfExistingNewsStory();
    testDeleteAnExistingNewsStory();
    testDeleteAnExistingNewsStoryIfItDoesnotExist();

    // IMPORTANT: Do this before calling filter test cases
    generateDataForFilterTestCases();

    testGetNewsByFilterTitle();
    testGetNewsByFilterAuthor();
    testGetNewsByFilterRangeDate();
    testGetNewsByFilterTitleAuthor();
    testGetNewsByFilterTitleAuthorRangeDate();

    if (allTestCasesPassed) console.log("All test cases passed");
    else console.log("Some test cases Failed!");
}

// Starts here!
driverTestCases();



loginPage.js
const header = `<head>
<meta charset="UTF-8">
<meta http-equiv="Cache-Control" content="no-cache, no-store, must-revalidate" />
<meta http-equiv="Pragma" content="no-cache" />
<meta http-equiv="Expires" content="0" />
<title>
    Activity2 pkedilay
</title>
</head>`;

function headerPanel(username, role) {
    return (`<div><p>Welcome ${username}, you are logged in as a ${role}</p></div>
                <div><a href = "/logout"> Logout </a></div>`);
}

function loginPage(username='', role='') {
    return (
        `<!DOCTYPE html>
        <html>
        
        ${header}
        
        <body>
            <form action="/login" method="POST">
                <div>
                    <label for="username">Username: </label>
                    <input type="text" name="username" value= "${username}" required />
                </div>
                <div>
                    <label for="password">Password: </label>
                    <input type="password" name="password" required />
                </div>
                <div>
                    
                    <input type="radio" name="role" value="author" ${isChecked('author', role)} />
                    <label for="author">Author </label>
                
                    
                    <input type="radio" name="role" value="guest" ${isChecked('guest', role)} />
                    <label for="guest">Guest </label>
                
                    
                    <input type="radio" name="role" value="subscriber" ${isChecked('subscriber', role)} />
                    <label for="subscriber" default>Subscriber </label>
                </div>
                <div>
                    <input type="submit" value="Log In" />
                </div>
            </form>
        </body>
        
        </html>`
    );
};

function isChecked(option, role) {
    if (option === role || (role === '' && option === 'guest')) {
        return 'checked';
    }
    return '';
}

function errorPage() {
    return (
        `<!DOCTYPE html>
        <html>
        
        ${header}
        
        <body>
            
            <div>
                <h1>
                    Username or Password Incorrect
                </h1><br />
                <h3>
                    <a href = "/">Click here to login again</a>
                </h3>
            </div>
        
        </body>
        
        </html>`
    )
}

module.exports = {loginPage, errorPage, header, headerPanel};

createNews.js
const header = require('./loginPage');

function createNews(username, role) {
    return (`
        <!DOCTYPE html>
        <html>
        ${header.header}
            <body>
            ${header.headerPanel(username, role)}
                <h2>Enter the story below</h2>
                <form action="/saveNews" method="POST">
                    <label>Title:</label> <input type="text" name="title" /><br />
                    <label>Content:</label> <br /><textarea name="storyContent" ></textarea><br />
                    <label>Public?</label> 
                    <input type="radio" name="isPrivate" value="false" checked />No
                    <input type="radio" name="isPrivate" value="true" />Yes
                    <br />
                    <label>Date: </label><input type="datetime-local" name="date" required />
                    <br />
                    <input type="submit" value="Save" />
                    
                </form>
                <form action="/dashboard" method="GET">
                    <button type="submit" >Cancel</button>
                </form>
            </body>
        </html>
    `)
}

module.exports ={ createNews };

NewsService.js
function newsStory (id, author, title, isPublic, storyContent, date) {
    
    this.id = id;
    this.author = author;
    this.title = title;
    this.isPublic = isPublic;
    this.storyContent = storyContent;
    this.date = date;
}

function newsStore () {

    this.news = [];
    this.index = 0;
    
    this.writeNewNewsStory = (author, title, isPublic, storyContent, date) => {
        try {
            this.news[this.index] = new newsStory(this.index++, author, title, isPublic, storyContent, date);
        } catch (error) {
            throw "Failed to write story, Please try again!";
        }
    }
    
    this.updateHeadlineOfNewsStory = (id, newTitle) => {
        try {
            this.news[id].title = newTitle;
        } catch (error) {
            throw `${id} does not exist`;
        }
    }
    
    this.changeContentOfExistingNewsStory = (id, author, title, isPublic, storyContent, date) => {
        try {
            this.news[id].author = author,
            this.news[id].title = title,
            this.news[id].isPublic = isPublic,
            this.news[id].storyContent = storyContent,
            this.news[id].date = date;
        } catch (error) {
            throw `${id} does not exist`;
        }
    }
    
    this.deleteAnExistingNewsStory = (id) => {
        try {
            if (this.news[id] === undefined) {
                throw error
            }
            delete this.news[id];
        } catch (error) {
            throw `${id} does not exist`;
        }
    }
    
    this.getNewsByFilter = (filterTitle, filterAuthor, startDate, endDate) => {
        var resultArray = [];
        for (let i in this.news) {
            if (filterAuthor !== undefined && this.news[i].author !== filterAuthor) {
                continue;
            }
            if (filterTitle !== undefined && !this.news[i].title.includes(filterTitle)) {
                continue;
            }
            if (startDate !== undefined && endDate !== undefined 
                && !(this.news[i].date.getTime() >= startDate.getTime() && this.news[i].date.getTime() <= endDate.getTime())) {
                    continue;
            }
            resultArray.push(this.news[i]);
        }
        return resultArray;
    }
}

module.exports = {newsStore}; // comment this if using .load

ExtraCredits/Logger.js
const events = require('events');
const fs = require('fs');

function Logger() {
    this.logs = "";

    events.EventEmitter.call(this);

    this.log = (message) => {
        this.logs += message + '\n';
    }
    this.on('userLogin', (message) => {
        this.log(`${new Date()} | ${message.ip} | INFO | User logged in: ${message.username}`);
    });

    this.on('userLogout', (message) => {
        this.log(`${new Date()} | ${message.ip} | INFO | User logged out: ${message.username}`);
    });

    this.on('userLoginFailed', (message) => {
        this.log(`${new Date()} | ${message.ip} | INFO | Failed login attempt: ${message.username}`);
    });

    this.on('errorOccured', (message) => {
        this.log(`${new Date()} | ${message.ip} | ERROR | Request failed for: ${message.username} \n ${message.error}`);
    });

    setInterval(() => {
        this.logs !== "" && this.writeFile(this.logs);
        this.logs = "";
    }, 5000);

    this.writeFile = (data) => {
        var options = {encoding:'utf8', flag:'a'};
        fs.writeFileSync('./activity2.log', data, options);
    }
}

Logger.prototype.__proto__ = events.EventEmitter.prototype;

module.exports = Logger;

ExtraCredits/viewNews.js
const header = require('./loginPage');

function viewNews(store, username, role) {
    return (
        `
        <!DOCTYPE html>
        <html>
        ${header.header}
            <body>
                ${header.headerPanel(username, role)}
                ${checkAuthor(role)}
                <hr>
                ${getNewsList(store, role, username).map(n => `${n} <br />`).join('')}
            </body>
        </html>
        `
    )
}

function checkAuthor(role) {
    if (role === 'author') {
        return `<br /><a href="/create">New Story</a>`;
    }
    return '';
}

function getNewsList(store, role, username) {
    let newsList = store.getNewsByFilter();
    if (role === 'subscriber') {
        newsList = newsList.map(element => {
            return element = `<a href="/news/${element.id}"> ${element.title} </a>`;
        });
    } else if (role === 'guest') {
        newsList = newsList.map(element => {
            if (element.isPublic) {
                return `<a href="/news/${element.id}"> ${element.title} </a>`;
            } else {
                return element.title;
            }
        });
    } else {
        newsList = newsList.map(element => {
            if (element.isPublic || element.author === username) {
                return `<a href="/news/${element.id}"> ${element.title} </a>`;
            } else {
                return element.title;
            }
        });
    }
    return newsList;
}

function renderNewsPage(response, news, author, retry, role) {
    if (news === undefined) {
        response.writeHead(404, {"Content-Type": "text/html"});
        response.write(`<h1>404: Page Not Found</h1>`);
        return;
    }
    if ((role === 'guest' && news.isPublic === false) 
        || (role === 'author' && news.author !== author && news.isPublic === false)) {
        response.writeHead(403, {"Content-Type": "text/html"});
        response.write(`<h1>403: Forbidden</h1>`);
        return;
    }
    response.writeHead(200, {
        'Content-type': 'text/html',
        "Cache-Control": "no-cache, no-store, must-revalidate"
    });
    response.write(`<!DOCTYPE html>
    <html>
    ${header.header}
        <body>
            ${header.headerPanel(author, role)}
            ${isRetry(retry)}
            <h2>${news.title}</h2>
            <i> Written by: ${news.author} </i> <br />
            <i> Published on ${news.date} </i>
            <br />
            
            <p>${news.storyContent}</p><br />

            ${deleteButton(news, author)}
        </body>
    </html>`)

}

function isRetry(retry) {
    if (retry === true) {
        return `<h2>Could not delete</h2>`;
    }
    return '';
}

function deleteButton(news, author) {
    if (news.author === author) {
        return `<a href='/delete/${news.id}'>Delete</a>`;
    }
    return ``;
}
module.exports = { viewNews, renderNewsPage };

activity2.js
const loginPage = require('./loginPage');
const viewNews = require('./viewNews');
const newsStore = require('./NewsService');
const createNews = require('./createNews');

const http = require('http');
const qs = require('querystring');
const url = require('url');

const httpHeader = {
    'Content-type': 'text/html', 
    "Cache-Control": "no-cache, no-store, must-revalidate"
}

const store = new newsStore.newsStore();
store.writeNewNewsStory('author', 'Software Engineering', true, 'This is a sample work', new Date(2000, 01, 01));
store.writeNewNewsStory('author', 'Web Development', false, 'This is a sample work', new Date(2020, 01, 01));
store.writeNewNewsStory('author1', 'Software Engineering', true, 'This is a sample work', new Date(2019, 01, 01));
store.writeNewNewsStory('author2', 'Node Development', true, 'This is a sample work', new Date(2018, 01, 01));
store.writeNewNewsStory('author2', 'Backend Development', true, 'This is a sample work', new Date(2018, 01, 01));

http.createServer(onRequest).listen(3000);

function onRequest(request, response) {
    try {
        var cookieList = parseCookies(request);
        var username = cookieList['username'];
        var role = cookieList['role'];
        var isLoggedIn = cookieList['isLoggedIn'];
        var requestUrl = url.parse(request.url).pathname;

        if (requestUrl === '/') {
            
            if (isLoggedIn === 'false') {
                response.writeHead(200, httpHeader);
                response.write(`<div><h1>Welcome ${role}: ${username}, Please enter your password</h1></div>
                 ${loginPage.loginPage(username, role)}`);
            } else if (username === undefined || role === undefined || isLoggedIn === undefined) {
                response.writeHead(200, httpHeader);
                response.write(loginPage.loginPage());
            } else {
                response.writeHead(301, {
                    'Content-type': 'text/html', 
                    "Cache-Control": "no-cache, no-store, must-revalidate",
                    Location: '/dashboard'
                });
            }
            response.end();
        } else if (requestUrl === '/login') {
            validateUser(request, response);
        } else if (isLoggedIn === 'true') {
            if (requestUrl === '/logout') {
                response.writeHead(200, {
                    'Set-Cookie': [`username=${username};`, `role=${role};`, `isLoggedIn=false;`],
                    'Content-type': 'text/html', 
                    "Cache-Control": "no-cache, no-store, must-revalidate"
                });
                response.write(`<div><h1>Welcome ${role}: ${username}, Please enter your password</h1></div>
                 ${loginPage.loginPage(username, role)}`);
                response.end();
            } else if (requestUrl.includes('/news/')) {
                let uri = requestUrl;
                let id = uri.substring(uri.lastIndexOf('/') + 1);
                viewNews.renderNewsPage(response, store.news[id], username, false, role);
                response.end();
            } else if (role==='author' && requestUrl === '/create') {
                response.writeHead(200,  httpHeader );
                response.write(createNews.createNews(username, role));
                response.end();
            } else if (role==='author' && requestUrl === '/saveNews') {
                saveNews(request, response, username, role);
            } else if (requestUrl === '/dashboard') {
                response.write(viewNews.viewNews(store, username, role));
                response.end();
            } else if (role==='author' && requestUrl.includes('/delete/')) {
                let uri = requestUrl;
                let id = uri.substring(uri.lastIndexOf('/') + 1);
                deleteStory(response, id, username, role);
            } else {
                response.writeHead(404, httpHeader);
                response.write(`<h1>404: Page Not Found</h1>`);
                response.end();
            }
        } else {
            response.writeHead(301, {
                'Content-type': 'text/html', 
                "Cache-Control": "no-cache, no-store, must-revalidate",
                Location: '/'
            });
            response.end();
        }
    } catch (e) {
        response.writeHead(500, httpHeader);
        response.write(`<h1>500: Internal Server Error</h1>`);
        response.end();
    }
}

function parseCookies (request) {
    var list = {},
        rc = request.headers.cookie;

    rc && rc.split(';').forEach(function( cookie ) {
        var parts = cookie.split('=');
        list[parts.shift().trim()] = decodeURI(parts.join('='));
    });

    return list;
}


function validateUser(request, response) {
    if (request.method == 'POST') {
        let body = '';
        request.on('data', function (data) {
            body += data;
        });

        request.on('end', function () {
            let post = qs.parse(body);
            username = post.username;
            role = post.role;
            if (userDetails[post.username] !== undefined
                && userDetails[post.username].password === post.password 
                && userDetails[post.username].role === post.role) {
                    response.writeHead(301, {
                        'Set-Cookie': [`username=${post.username};`, `role=${post.role};`, `isLoggedIn=true;`],
                        'Content-type': 'text/html', "Cache-Control": "no-cache, no-store, must-revalidate",
                        Location: '/dashboard'
                    });
                // /response.writeHead(301, { Location: '/dashboard' } );
            } else {
                response.writeHead(200, httpHeader);
                response.write(loginPage.errorPage());
            }
            response.end();
        });
    } else {
        response.writeHead(405, httpHeader);
        response.write(`<h2>405 Method Not Allowed</h2>`);
        response.end();
    }
}

function saveNews(request, response, username, role) {
    if (request.method == 'POST') {
        try {
            let body = '';
            request.on('data', function (data) {
                body += data;
            });

            request.on('end', function () {
                let post = qs.parse(body);
                if (post !== undefined) {
                    store.writeNewNewsStory(username, post.title, (post.isPrivate === 'true'), post.storyContent, new Date(post.date));
                    response.writeHead(301, {
                        'Content-type': 'text/html', 
                        "Cache-Control": "no-cache, no-store, must-revalidate",
                        Location: '/dashboard'
                    });
                    response.end();
                } else {
                    response.writeHead(500, httpHeader);
                    response.write(`
                        <h2>Internal Server Error: Could not save the data. Try again!</h2>
                        ${createNews.createNews(username, role)}
                    `);
                    response.end();
                }
            });
            
        } catch (e) {
            response.writeHead(500, httpHeader);
            response.write(`
                <h2>Internal Server Error: Could not save the data. Try again!</h2>
                ${createNews.createNews(username, role)}
            `);
            response.end();
        }
    } else {
        response.writeHead(405, httpHeader);
        response.write(`<h2>405 Method Not Allowed</h2>`);
        response.end();
    }
}

function deleteStory(response, id, username, role) {
    try {
        if (store.news[id] !== undefined && store.news[id].author === username) {
            store.deleteAnExistingNewsStory(id);
        }
        
        response.writeHead(301, { 
            'Content-type': 'text/html', 
            "Cache-Control": "no-cache, no-store, must-revalidate",
            Location: '/dashboard'
        });
        response.end();
    } catch (e) {
        response.writeHead(500, httpHeader);
        viewNews.renderNewsPage(response, store.news[id], username, true, role);
        response.end();
    }
}

const userDetails = {
    'author': {
        username: "author",
        role: "author",
        password: "author"
    },
    'guest': {
        username: "guest",
        role: "guest",
        password: "guest"
    },
    'subscriber': {
        username: "subscriber",
        role: "subscriber",
        password: "subscriber"
    },
    'author1': {
        username: "author1",
        role: "author",
        password: "author1"
    },
    'author2': {
        username: "author2",
        role: "author",
        password: "author2"
    }
}

